File: Circular_Linked_list/circular_doubly_linked_list_delete_operations.py
# File Name: circular_doubly_linked_list_delete_operations.py
# Program to delete a node at beginning, end, and given position
# in a circular doubly linked list

class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

head = None

# Delete at beginning
def delete_beg():
    global head
    if head is None:
        print("List is empty")
        return

    if head.next == head:
        head = None
        return

    last = head.prev
    head = head.next
    head.prev = last
    last.next = head

# Delete at end
def delete_end():
    global head
    if head is None:
        print("List is empty")
        return

    if head.next == head:
        head = None
        return

    last = head.prev
    second_last = last.prev
    second_last.next = head
    head.prev = second_last

# Delete at given position (1-based)
def delete_pos(pos):
    global head
    if head is None:
        print("List is empty")
        return

    if pos == 1:
        delete_beg()
        return

    temp = head
    i = 1
    while i < pos and temp.next != head:
        temp = temp.next
        i += 1

    if i != pos:
        print("Invalid position")
        return

    temp.prev.next = temp.next
    temp.next.prev = temp.prev

if __name__ == "__main__":
    delete_beg()
    delete_end()
    delete_pos(2)

File: Circular_Linked_list/CircularDoublyLinkedListDeleteOperations.java
// File Name: CircularDoublyLinkedListDeleteOperations.java
// Program to delete a node at beginning, end, and given position
// in a circular doubly linked list

class Node {
    int data;
    Node prev;
    Node next;

    Node(int data) {
        this.data = data;
    }
}

public class CircularDoublyLinkedListDeleteOperations {
    static Node head = null;

    // Delete at beginning
    static void deleteBeg() {
        if (head == null) {
            System.out.println("List is empty");
            return;
        }

        if (head.next == head) {
            head = null;
            return;
        }

        Node last = head.prev;
        head = head.next;
        head.prev = last;
        last.next = head;
    }

    // Delete at end
    static void deleteEnd() {
        if (head == null) {
            System.out.println("List is empty");
            return;
        }

        if (head.next == head) {
            head = null;
            return;
        }

        Node last = head.prev;
        Node secondLast = last.prev;

        secondLast.next = head;
        head.prev = secondLast;
    }

    // Delete at given position (1-based)
    static void deletePos(int pos) {
        if (head == null) {
            System.out.println("List is empty");
            return;
        }

        if (pos == 1) {
            deleteBeg();
            return;
        }

        Node temp = head;
        int i;

        for (i = 1; i < pos && temp.next != head; i++)
            temp = temp.next;

        if (i != pos) {
            System.out.println("Invalid position");
            return;
        }

        temp.prev.next = temp.next;
        temp.next.prev = temp.prev;
    }

    public static void main(String[] args) {
        deleteBeg();
        deleteEnd();
        deletePos(2);
    }
}
